File: packages/analyze/src/claude-md.ts
import { stat, readFile } from 'node:fs/promises';
import * as path from 'node:path';

import type { TurnRecord } from '@relayburn/reader';

import { lookupModelRate } from './cost.js';
import type { PricingTable } from './pricing.js';

const PER_MILLION = 1_000_000;
const CHARS_PER_TOKEN = 4;

export interface MarkdownSection {
  heading: string;
  level: number; // 0 for preamble, 1-6 for # through ######
  startLine: number; // 1-indexed
  endLine: number; // 1-indexed inclusive
  bytes: number;
  tokens: number;
}

export interface ParsedClaudeMd {
  path: string;
  totalLines: number;
  bytes: number;
  tokens: number;
  // Top-level sections at the inferred grouping level (H2 if any exist,
  // otherwise H1, otherwise just preamble).
  sections: MarkdownSection[];
  groupingLevel: number; // 1 or 2; 0 if no headings
}

export interface SessionClaudeMdCost {
  sessionId: string;
  cost: number;
  ridingTurns: number;
  totalTurns: number;
  model: string;
}

export interface SectionCost {
  filePath: string;
  section: MarkdownSection;
  tokenShare: number; // section.tokens / total tokens across all files
  costPerSession: number; // average per-session cost attributed to this section
  totalCost: number; // sum across all sessions
}

export interface ClaudeMdAttributionResult {
  totalTokens: number;
  totalCost: number;
  sessionCosts: SessionClaudeMdCost[];
  sectionCosts: SectionCost[];
  perSessionAvg: number;
  perSessionP95: number;
  sessionCount: number;
}

export interface AttributeClaudeMdInput {
  files: ParsedClaudeMd[];
  turns: TurnRecord[];
  pricing: PricingTable;
}

export async function findClaudeMdFiles(projectPath: string): Promise<string[]> {
  const candidates = [
    path.join(projectPath, 'CLAUDE.md'),
    path.join(projectPath, '.claude', 'CLAUDE.md'),
  ];
  const found: string[] = [];
  for (const c of candidates) {
    try {
      const st = await stat(c);
      if (st.isFile()) found.push(c);
    } catch {
      // ignore missing
    }
  }
  return found;
}

export async function loadClaudeMdFile(filePath: string): Promise<ParsedClaudeMd> {
  const text = await readFile(filePath, 'utf8');
  return parseClaudeMd(filePath, text);
}

export function parseClaudeMd(filePath: string, text: string): ParsedClaudeMd {
  // Normalize CRLF → LF and drop a single trailing newline so `totalLines`
  // and per-section `endLine` match what a user sees in an editor. Empty text
  // => 0 lines.
  const normalized = text.replace(/\r\n/g, '\n');
  const trimmedEnd = normalized.endsWith('\n')
    ? normalized.slice(0, -1)
    : normalized;
  const lines = trimmedEnd.length === 0 ? [] : trimmedEnd.split('\n');
  const totalLines = lines.length;
  // All byte-counts downstream are derived from `normalized` so section bytes
  // are additive (Σ section.bytes == totalBytes exactly). This avoids
  // non-additive ceil() rounding when computing per-section shares.
  const totalBytes = Buffer.byteLength(normalized, 'utf8');
  const tokens = Math.max(0, Math.ceil(totalBytes / CHARS_PER_TOKEN));

  // Precompute per-line byte lengths (excluding the trailing \n); the
  // "weight" of line i on disk is lineBytes[i] + 1 for the newline, except
  // the last line if the normalized text has no trailing newline.
  const lineBytes = lines.map((l) => Buffer.byteLength(l, 'utf8'));
  const hadTrailingNewline = normalized.endsWith('\n');
  const lineWithNewlineWeight = (idx: number): number => {
    const base = lineBytes[idx] ?? 0;
    const isLast = idx === totalLines - 1;
    if (isLast && !hadTrailingNewline) return base;
    return base + 1;
  };
  const rangeBytes = (startLine1: number, endLine1: number): number => {
    let sum = 0;
    for (let i = startLine1 - 1; i <= endLine1 - 1; i++) sum += lineWithNewlineWeight(i);
    return sum;
  };

  const headings = findHeadings(lines);
  // Choose the grouping level:
  //   - If any H2 exists: group at H2 (per the issue's example output).
  //   - Else if any H1 exists: group at H1.
  //   - Else: no headings at all — single preamble section.
  let groupingLevel = 0;
  if (headings.some((h) => h.level === 2)) groupingLevel = 2;
  else if (headings.some((h) => h.level === 1)) groupingLevel = 1;

  const sections: MarkdownSection[] = [];
  if (groupingLevel === 0) {
    if (totalLines > 0 && totalBytes > 0) {
      sections.push({
        heading: '(preamble)',
        level: 0,
        startLine: 1,
        endLine: totalLines,
        bytes: totalBytes,
        tokens,
      });
    }
    return { path: filePath, totalLines, bytes: totalBytes, tokens, sections, groupingLevel };
  }

  // Only headings AT the grouping level become top-level sections. Higher-
  // level headings (e.g. an H1 doc title above a series of H2 sections) get
  // folded into the preamble, matching the standard CLAUDE.md shape where
  // the H1 is a document title and H2s are the meaningful sections.
  const groupHeadings = headings.filter((h) => h.level === groupingLevel);
  // Preamble: content before the first grouping heading.
  const firstStart = groupHeadings[0]?.line ?? totalLines + 1;
  if (firstStart > 1) {
    const pbBytes = rangeBytes(1, firstStart - 1);
    if (pbBytes > 0) {
      sections.push({
        heading: '(preamble)',
        level: 0,
        startLine: 1,
        endLine: firstStart - 1,
        bytes: pbBytes,
        tokens: Math.max(0, Math.ceil(pbBytes / CHARS_PER_TOKEN)),
      });
    }
  }

  for (let i = 0; i < groupHeadings.length; i++) {
    const h = groupHeadings[i]!;
    const next = groupHeadings[i + 1];
    const endLine = next ? next.line - 1 : totalLines;
    const secBytes = rangeBytes(h.line, endLine);
    sections.push({
      heading: h.text,
      level: h.level,
      startLine: h.line,
      endLine,
      bytes: secBytes,
      tokens: Math.max(0, Math.ceil(secBytes / CHARS_PER_TOKEN)),
    });
  }
  return { path: filePath, totalLines, bytes: totalBytes, tokens, sections, groupingLevel };
}

interface HeadingInfo {
  line: number; // 1-indexed
  level: number;
  text: string; // includes the leading hashes for display
}

function findHeadings(lines: string[]): HeadingInfo[] {
  const out: HeadingInfo[] = [];
  let fenceChar = ''; // '`' or '~' while inside a fence, '' otherwise
  let fenceLen = 0; // length of the opening fence run
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]!;
    const trimmed = line.trim();
    if (fenceChar === '') {
      // Opening fence: line starts with 3+ backticks or tildes.
      const open = trimmed.match(/^(`{3,}|~{3,})/);
      if (open) {
        fenceChar = open[1]!.charAt(0);
        fenceLen = open[1]!.length;
      }
      // An info string after the opening run is allowed, so we keep scanning
      // only if we're already inside a fence.
      if (fenceChar !== '') continue;
    } else {
      // Closing fence per CommonMark: a run of the same fence character, at
      // least as long as the opening run, followed only by whitespace. Lines
      // that start with fence chars but have other trailing content (e.g.
      // "```python" nested inside a 3-backtick block) do NOT close.
      const closeRe = new RegExp(`^${fenceChar === '`' ? '`' : '~'}{${fenceLen},}\\s*$`);
      if (closeRe.test(trimmed)) {
        fenceChar = '';
        fenceLen = 0;
      }
      continue;
    }
    const m = line.match(/^(#{1,6})\s+(.*\S)\s*$/);
    if (m) {
      const hashes = m[1]!;
      out.push({ line: i + 1, level: hashes.length, text: `${hashes} ${m[2]}` });
    }
  }
  return out;
}

export function attributeClaudeMd(
  input: AttributeClaudeMdInput,
): ClaudeMdAttributionResult {
  const totalTokens = input.files.reduce((sum, f) => sum + f.tokens, 0);
  if (totalTokens === 0) {
    return {
      totalTokens: 0,
      totalCost: 0,
      sessionCosts: [],
      sectionCosts: [],
      perSessionAvg: 0,
      perSessionP95: 0,
      sessionCount: 0,
    };
  }

  const bySession = new Map<string, TurnRecord[]>();
  for (const t of input.turns) {
    let list = bySession.get(t.sessionId);
    if (!list) {
      list = [];
      bySession.set(t.sessionId, list);
    }
    list.push(t);
  }

  const sessionCosts: SessionClaudeMdCost[] = [];
  let totalCost = 0;
  for (const [sessionId, turns] of bySession) {
    turns.sort((a, b) => a.turnIndex - b.turnIndex);
    let cost = 0;
    let ridingTurns = 0;
    const modelCounts = new Map<string, number>();
    for (const t of turns) {
      const rate = lookupModelRate(t.model, input.pricing);
      if (!rate) continue;
      modelCounts.set(t.model, (modelCounts.get(t.model) ?? 0) + 1);
      // Treat CLAUDE.md as residing in cache once any turn reads enough cached
      // tokens to fit it. This is conservative: if a turn's cacheRead is below
      // claude_md_tokens, the file may have been compacted away or this is the
      // first turn (CLAUDE.md is in cacheCreate that turn, not cacheRead).
      if (t.usage.cacheRead < totalTokens) continue;
      cost += (totalTokens / PER_MILLION) * rate.cacheRead;
      ridingTurns++;
    }
    // Record every session in the query window — including those with zero
    // attributed cost — so `perSessionAvg`/`perSessionP95`/`sessionCount`
    // reflect the whole window, not only sessions where CLAUDE.md was cached.
    const dominantModel = pickDominantModel(modelCounts);
    sessionCosts.push({
      sessionId,
      cost,
      ridingTurns,
      totalTurns: turns.length,
      model: dominantModel,
    });
    totalCost += cost;
  }

  const sessionCostValues = sessionCosts.map((s) => s.cost).sort((a, b) => a - b);
  const perSessionAvg = sessionCostValues.length === 0
    ? 0
    : sessionCostValues.reduce((a, b) => a + b, 0) / sessionCostValues.length;
  const perSessionP95 = percentile(sessionCostValues, 0.95);

  // Use bytes (additive) rather than per-section token counts (each ceil()ed
  // independently, so they can sum to more than `totalTokens`) for the share.
  // This keeps Σ sectionCost ≤ totalCost exactly.
  const totalBytes = input.files.reduce((sum, f) => sum + f.bytes, 0);
  const sectionCosts: SectionCost[] = [];
  for (const f of input.files) {
    for (const section of f.sections) {
      const tokenShare = totalBytes > 0 ? section.bytes / totalBytes : 0;
      const totalSecCost = totalCost * tokenShare;
      const perSessionSecCost = perSessionAvg * tokenShare;
      sectionCosts.push({
        filePath: f.path,
        section,
        tokenShare,
        costPerSession: perSessionSecCost,
        totalCost: totalSecCost,
      });
    }
  }
  sectionCosts.sort((a, b) => b.totalCost - a.totalCost);

  return {
    totalTokens,
    totalCost,
    sessionCosts,
    sectionCosts,
    perSessionAvg,
    perSessionP95,
    sessionCount: sessionCosts.length,
  };
}

function pickDominantModel(counts: Map<string, number>): string {
  let bestModel = '';
  let bestCount = -1;
  for (const [m, c] of counts) {
    if (c > bestCount) {
      bestModel = m;
      bestCount = c;
    }
  }
  return bestModel;
}

function percentile(sorted: number[], p: number): number {
  if (sorted.length === 0) return 0;
  if (sorted.length === 1) return sorted[0]!;
  const idx = Math.min(sorted.length - 1, Math.max(0, Math.ceil(p * sorted.length) - 1));
  return sorted[idx]!;
}

export interface TrimRecommendation {
  filePath: string;
  section: MarkdownSection;
  projectedSavingsPerSession: number;
  projectedSavingsAcrossWindow: number;
}

export function buildTrimRecommendations(
  attribution: ClaudeMdAttributionResult,
  topN = 3,
): TrimRecommendation[] {
  // Pick the most expensive non-preamble sections as TRIM candidates.
  const candidates = attribution.sectionCosts.filter((s) => s.section.level > 0);
  const top = candidates.slice(0, topN);
  return top.map((s) => ({
    filePath: s.filePath,
    section: s.section,
    projectedSavingsPerSession: s.costPerSession,
    projectedSavingsAcrossWindow: s.totalCost,
  }));
}

export function renderUnifiedDiffForRecommendation(
  filePath: string,
  fileText: string,
  rec: TrimRecommendation,
  baseDir?: string,
): string {
  const normalized = fileText.replace(/\r\n/g, '\n');
  const trimmedEnd = normalized.endsWith('\n') ? normalized.slice(0, -1) : normalized;
  const lines = trimmedEnd.length === 0 ? [] : trimmedEnd.split('\n');
  const start = rec.section.startLine;
  const end = rec.section.endLine;
  const removed = lines.slice(start - 1, end);
  const display = toPosixRelative(filePath, baseDir);
  const header = [`--- a/${display}`, `+++ b/${display}`];
  // Hunk header: original = start, count = removed.length; new = start, count = 0
  const hunk = `@@ -${start},${removed.length} +${start},0 @@`;
  const body = removed.map((l) => `-${l}`).join('\n');
  return [
    `# TRIM: ${rec.section.heading}`,
    `# projected savings per session: $${rec.projectedSavingsPerSession.toFixed(4)}`,
    `# projected savings across window: $${rec.projectedSavingsAcrossWindow.toFixed(4)}`,
    ...header,
    hunk,
    body,
  ].join('\n');
}

function toPosixRelative(filePath: string, baseDir?: string): string {
  let p = filePath;
  if (baseDir) {
    const rel = path.relative(baseDir, filePath);
    if (rel && !rel.startsWith('..')) p = rel;
  }
  // Convert Windows separators to POSIX for diff-friendly paths, and strip
  // a leading separator so diff headers aren't `--- a//abs/path`.
  p = p.split(path.sep).join('/').replace(/^\/+/, '');
  return p;
}

File: crates/relayburn-analyze/src/claude_md.rs
//! Markdown section parser + per-session attribution for `CLAUDE.md` /
//! `AGENTS.md` overhead files. Rust port of `packages/analyze/src/claude-md.ts`.
//!
//! This module is the parsing/attribution kernel that the higher-level
//! `overhead` module composes when answering "how much of this session's
//! spend went to system overhead context loaded into every prompt?". Math is
//! `f64` and accumulation order matches the TS reduce loops so the
//! per-session/per-section USD totals stay within the 1e-9 USD precision
//! contract the parent issue gates on.
//!
//! See AgentWorkforce/burn#272 (claude-md attribution) and #276 (overhead).

use std::fs;
use std::path::{Path, PathBuf};

use indexmap::IndexMap;
use regex::Regex;
use relayburn_reader::TurnRecord;

use crate::cost::lookup_model_rate;
use crate::pricing::PricingTable;

const PER_MILLION: f64 = 1_000_000.0;
const CHARS_PER_TOKEN: u64 = 4;

#[derive(Debug, Clone, PartialEq)]
pub struct MarkdownSection {
    pub heading: String,
    /// 0 for preamble, 1-6 for `#` through `######`.
    pub level: u32,
    pub start_line: u32,
    pub end_line: u32,
    pub bytes: u64,
    pub tokens: u64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ParsedClaudeMd {
    pub path: String,
    pub total_lines: u32,
    pub bytes: u64,
    pub tokens: u64,
    pub sections: Vec<MarkdownSection>,
    /// 1 or 2 (the heading depth at which sections were grouped); 0 if no
    /// headings exist at all.
    pub grouping_level: u32,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SessionClaudeMdCost {
    pub session_id: String,
    pub cost: f64,
    pub riding_turns: u64,
    pub total_turns: u64,
    pub model: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SectionCost {
    pub file_path: String,
    pub section: MarkdownSection,
    pub token_share: f64,
    pub cost_per_session: f64,
    pub total_cost: f64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ClaudeMdAttributionResult {
    pub total_tokens: u64,
    pub total_cost: f64,
    pub session_costs: Vec<SessionClaudeMdCost>,
    pub section_costs: Vec<SectionCost>,
    pub per_session_avg: f64,
    pub per_session_p95: f64,
    pub session_count: u64,
}

pub struct AttributeClaudeMdInput<'a> {
    pub files: &'a [ParsedClaudeMd],
    pub turns: &'a [TurnRecord],
    pub pricing: &'a PricingTable,
}

pub fn find_claude_md_files(project_path: &Path) -> Vec<PathBuf> {
    let candidates = [
        project_path.join("CLAUDE.md"),
        project_path.join(".claude").join("CLAUDE.md"),
    ];
    candidates
        .into_iter()
        .filter(|p| fs::metadata(p).map(|m| m.is_file()).unwrap_or(false))
        .collect()
}

pub fn load_claude_md_file(file_path: &Path) -> std::io::Result<ParsedClaudeMd> {
    let text = fs::read_to_string(file_path)?;
    Ok(parse_claude_md(&file_path.to_string_lossy(), &text))
}

pub fn parse_claude_md(file_path: &str, text: &str) -> ParsedClaudeMd {
    // Normalize CRLF → LF and drop a single trailing newline so `total_lines`
    // and per-section `end_line` match what a user sees in an editor.
    let normalized = text.replace("\r\n", "\n");
    let had_trailing_newline = normalized.ends_with('\n');
    let trimmed_end: &str = if had_trailing_newline {
        &normalized[..normalized.len() - 1]
    } else {
        &normalized
    };
    let lines: Vec<&str> = if trimmed_end.is_empty() {
        Vec::new()
    } else {
        trimmed_end.split('\n').collect()
    };
    let total_lines = lines.len() as u32;
    let total_bytes = normalized.len() as u64;
    let tokens = ceil_div(total_bytes, CHARS_PER_TOKEN);

    let line_bytes: Vec<u64> = lines.iter().map(|l| l.len() as u64).collect();

    let line_with_newline_weight = |idx: usize| -> u64 {
        let base = line_bytes.get(idx).copied().unwrap_or(0);
        let is_last = idx + 1 == lines.len();
        if is_last && !had_trailing_newline {
            base
        } else {
            base + 1
        }
    };
    let range_bytes = |start_line1: u32, end_line1: u32| -> u64 {
        let mut sum = 0u64;
        for i in start_line1..=end_line1 {
            sum += line_with_newline_weight((i - 1) as usize);
        }
        sum
    };

    let headings = find_headings(&lines);
    let mut grouping_level = 0u32;
    if headings.iter().any(|h| h.level == 2) {
        grouping_level = 2;
    } else if headings.iter().any(|h| h.level == 1) {
        grouping_level = 1;
    }

    let mut sections: Vec<MarkdownSection> = Vec::new();
    if grouping_level == 0 {
        if total_lines > 0 && total_bytes > 0 {
            sections.push(MarkdownSection {
                heading: "(preamble)".to_string(),
                level: 0,
                start_line: 1,
                end_line: total_lines,
                bytes: total_bytes,
                tokens,
            });
        }
        return ParsedClaudeMd {
            path: file_path.to_string(),
            total_lines,
            bytes: total_bytes,
            tokens,
            sections,
            grouping_level,
        };
    }

    let group_headings: Vec<&HeadingInfo> = headings
        .iter()
        .filter(|h| h.level == grouping_level)
        .collect();
    let first_start = group_headings
        .first()
        .map(|h| h.line)
        .unwrap_or(total_lines + 1);
    if first_start > 1 {
        let pb_bytes = range_bytes(1, first_start - 1);
        if pb_bytes > 0 {
            sections.push(MarkdownSection {
                heading: "(preamble)".to_string(),
                level: 0,
                start_line: 1,
                end_line: first_start - 1,
                bytes: pb_bytes,
                tokens: ceil_div(pb_bytes, CHARS_PER_TOKEN),
            });
        }
    }
    for (i, h) in group_headings.iter().enumerate() {
        let next = group_headings.get(i + 1);
        let end_line = next.map(|n| n.line - 1).unwrap_or(total_lines);
        let sec_bytes = range_bytes(h.line, end_line);
        sections.push(MarkdownSection {
            heading: h.text.clone(),
            level: h.level,
            start_line: h.line,
            end_line,
            bytes: sec_bytes,
            tokens: ceil_div(sec_bytes, CHARS_PER_TOKEN),
        });
    }

    ParsedClaudeMd {
        path: file_path.to_string(),
        total_lines,
        bytes: total_bytes,
        tokens,
        sections,
        grouping_level,
    }
}

struct HeadingInfo {
    line: u32,
    level: u32,
    /// Includes the leading hashes for display (matches TS shape).
    text: String,
}

fn find_headings(lines: &[&str]) -> Vec<HeadingInfo> {
    // Heading: 1-6 hashes, whitespace, then text ending with a non-space char.
    let heading_re = Regex::new(r"^(#{1,6})\s+(.*\S)\s*$").unwrap();
    // Opening fence: 3+ backticks or tildes (after trim).
    let open_re = Regex::new(r"^(`{3,}|~{3,})").unwrap();

    let mut out: Vec<HeadingInfo> = Vec::new();
    let mut fence_char: Option<char> = None;
    let mut fence_len: usize = 0;

    for (i, line) in lines.iter().enumerate() {
        let trimmed = line.trim();
        if fence_char.is_none() {
            if let Some(caps) = open_re.captures(trimmed) {
                let m = caps.get(1).unwrap().as_str();
                fence_char = m.chars().next();
                fence_len = m.len();
                continue;
            }
        } else {
            let ch = fence_char.unwrap();
            // Closing fence per CommonMark: a run of the same fence char at
            // least as long as the opening run, followed only by whitespace.
            let mut closes = false;
            let s = trimmed;
            let mut run = 0usize;
            for c in s.chars() {
                if c == ch {
                    run += 1;
                } else {
                    break;
                }
            }
            if run >= fence_len {
                let rest = &s[run..];
                if rest.chars().all(|c| c.is_whitespace()) {
                    closes = true;
                }
            }
            if closes {
                fence_char = None;
                fence_len = 0;
            }
            continue;
        }
        if let Some(caps) = heading_re.captures(line) {
            let hashes = caps.get(1).unwrap().as_str();
            let body = caps.get(2).unwrap().as_str();
            out.push(HeadingInfo {
                line: (i + 1) as u32,
                level: hashes.len() as u32,
                text: format!("{} {}", hashes, body),
            });
        }
    }
    out
}

pub fn attribute_claude_md(input: AttributeClaudeMdInput<'_>) -> ClaudeMdAttributionResult {
    let total_tokens: u64 = input.files.iter().map(|f| f.tokens).sum();
    if total_tokens == 0 {
        return ClaudeMdAttributionResult {
            total_tokens: 0,
            total_cost: 0.0,
            session_costs: Vec::new(),
            section_costs: Vec::new(),
            per_session_avg: 0.0,
            per_session_p95: 0.0,
            session_count: 0,
        };
    }

    // Insertion-order-preserving group by sessionId — matches TS `Map`
    // iteration order, so downstream stable sorts tie-break consistently.
    let mut by_session: IndexMap<String, Vec<TurnRecord>> = IndexMap::new();
    for t in input.turns {
        by_session
            .entry(t.session_id.clone())
            .or_default()
            .push(t.clone());
    }

    let mut session_costs: Vec<SessionClaudeMdCost> = Vec::new();
    let mut total_cost = 0.0_f64;
    for (session_id, mut turns) in by_session.into_iter() {
        turns.sort_by_key(|t| t.turn_index);
        let mut cost = 0.0_f64;
        let mut riding_turns: u64 = 0;
        let mut model_counts: IndexMap<String, u64> = IndexMap::new();
        for t in &turns {
            let rate = match lookup_model_rate(&t.model, input.pricing) {
                Some(r) => r,
                None => continue,
            };
            *model_counts.entry(t.model.clone()).or_insert(0) += 1;
            // Treat the file as cache-resident only once a turn reads at least
            // `total_tokens` cached tokens — conservative match for the TS
            // `cacheRead < totalTokens` early-continue.
            if t.usage.cache_read < total_tokens {
                continue;
            }
            cost += (total_tokens as f64 / PER_MILLION) * rate.cache_read;
            riding_turns += 1;
        }
        let dominant_model = pick_dominant_model(&model_counts);
        session_costs.push(SessionClaudeMdCost {
            session_id,
            cost,
            riding_turns,
            total_turns: turns.len() as u64,
            model: dominant_model,
        });
        total_cost += cost;
    }

    let mut sorted_session_costs: Vec<f64> = session_costs.iter().map(|s| s.cost).collect();
    sorted_session_costs.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let per_session_avg = if sorted_session_costs.is_empty() {
        0.0
    } else {
        sorted_session_costs.iter().sum::<f64>() / sorted_session_costs.len() as f64
    };
    let per_session_p95 = percentile(&sorted_session_costs, 0.95);

    // Use bytes (additive) rather than per-section token counts (each ceil()ed
    // independently, so they can sum to more than `total_tokens`) for the
    // share. This keeps Σ section_cost ≤ total_cost exactly.
    let total_bytes: u64 = input.files.iter().map(|f| f.bytes).sum();
    let mut section_costs: Vec<SectionCost> = Vec::new();
    for f in input.files {
        for section in &f.sections {
            let token_share = if total_bytes > 0 {
                section.bytes as f64 / total_bytes as f64
            } else {
                0.0
            };
            let total_sec_cost = total_cost * token_share;
            let per_session_sec_cost = per_session_avg * token_share;
            section_costs.push(SectionCost {
                file_path: f.path.clone(),
                section: section.clone(),
                token_share,
                cost_per_session: per_session_sec_cost,
                total_cost: total_sec_cost,
            });
        }
    }
    section_costs.sort_by(|a, b| {
        b.total_cost
            .partial_cmp(&a.total_cost)
            .unwrap_or(std::cmp::Ordering::Equal)
    });

    let session_count = session_costs.len() as u64;
    ClaudeMdAttributionResult {
        total_tokens,
        total_cost,
        session_costs,
        section_costs,
        per_session_avg,
        per_session_p95,
        session_count,
    }
}

fn pick_dominant_model(counts: &IndexMap<String, u64>) -> String {
    let mut best_model = String::new();
    // TS uses `bestCount = -1` as a sentinel; matched by `Option<u64>` here.
    let mut best_count: Option<u64> = None;
    for (m, c) in counts {
        if best_count.is_none_or(|bc| *c > bc) {
            best_model = m.clone();
            best_count = Some(*c);
        }
    }
    best_model
}

fn percentile(sorted: &[f64], p: f64) -> f64 {
    if sorted.is_empty() {
        return 0.0;
    }
    if sorted.len() == 1 {
        return sorted[0];
    }
    let idx_f = (p * sorted.len() as f64).ceil() - 1.0;
    let idx = idx_f.max(0.0).min((sorted.len() - 1) as f64) as usize;
    sorted[idx]
}

fn ceil_div(n: u64, d: u64) -> u64 {
    if n == 0 {
        0
    } else {
        n.div_ceil(d)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use relayburn_reader::{SourceKind, Usage};

    fn turn(session: &str, idx: u64, model: &str, cache_read: u64) -> TurnRecord {
        TurnRecord {
            v: 1,
            source: SourceKind::ClaudeCode,
            session_id: session.to_string(),
            session_path: None,
            message_id: format!("m-{}", idx),
            turn_index: idx,
            ts: "2026-04-23T00:00:00.000Z".to_string(),
            model: model.to_string(),
            project: None,
            project_key: None,
            usage: Usage {
                input: 0,
                output: 0,
                reasoning: 0,
                cache_read,
                cache_create_5m: 0,
                cache_create_1h: 0,
            },
            tool_calls: Vec::new(),
            files_touched: None,
            subagent: None,
            stop_reason: None,
            activity: None,
            retries: None,
            has_edits: None,
            fidelity: None,
        }
    }

    #[test]
    fn parses_no_heading_file_as_single_preamble() {
        let parsed = parse_claude_md("/p/C.md", "just some prose\nover two lines\n");
        assert_eq!(parsed.grouping_level, 0);
        assert_eq!(parsed.sections.len(), 1);
        assert_eq!(parsed.sections[0].heading, "(preamble)");
        assert_eq!(parsed.total_lines, 2);
        // Σ section bytes == total bytes (preamble covers the whole file).
        assert_eq!(parsed.sections[0].bytes, parsed.bytes);
    }

    #[test]
    fn groups_at_h2_when_h2_exists() {
        let text = "# Title\n\n## Alpha\nbody-a\n## Beta\nbody-b\n";
        let parsed = parse_claude_md("/p/C.md", text);
        assert_eq!(parsed.grouping_level, 2);
        // preamble (Title) + Alpha + Beta
        assert_eq!(parsed.sections.len(), 3);
        assert_eq!(parsed.sections[0].heading, "(preamble)");
        assert_eq!(parsed.sections[1].heading, "## Alpha");
        assert_eq!(parsed.sections[2].heading, "## Beta");
        // Σ section.bytes == file.bytes (additivity invariant the attribution
        // arithmetic relies on for the 1e-9 USD precision contract).
        let sum: u64 = parsed.sections.iter().map(|s| s.bytes).sum();
        assert_eq!(sum, parsed.bytes);
    }

    #[test]
    fn skips_headings_inside_fenced_code_blocks() {
        let text = "## Real\nbody\n```\n## Fake\n```\n## Also Real\n";
        let parsed = parse_claude_md("/p/C.md", text);
        assert_eq!(parsed.grouping_level, 2);
        let names: Vec<_> = parsed
            .sections
            .iter()
            .map(|s| s.heading.as_str())
            .collect();
        assert_eq!(names, vec!["## Real", "## Also Real"]);
    }

    #[test]
    fn attribution_zero_when_total_tokens_zero() {
        let pricing = PricingTable::new();
        let parsed = ParsedClaudeMd {
            path: "/p/C.md".to_string(),
            total_lines: 0,
            bytes: 0,
            tokens: 0,
            sections: Vec::new(),
            grouping_level: 0,
        };
        let turns = vec![turn("s", 0, "claude-sonnet-4-6", 1000)];
        let result = attribute_claude_md(AttributeClaudeMdInput {
            files: &[parsed],
            turns: &turns,
            pricing: &pricing,
        });
        assert_eq!(result.total_tokens, 0);
        assert_eq!(result.total_cost, 0.0);
        assert_eq!(result.session_count, 0);
    }
}
